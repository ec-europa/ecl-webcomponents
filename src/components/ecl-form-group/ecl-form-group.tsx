import { Component, Prop, h } from '@stencil/core';

@Component({
  tag: 'ecl-form-group',
  styleUrls: {
    ec: './build/styles/ecl-form-group-ec.css',
    eu: './build/styles/ecl-form-group-eu.css',
  },
  shadow: false,
  scoped: true,
})

export class EclFormGroup {
  @Prop() styleClass: string = '';
  @Prop() theme: string = 'ec';
  @Prop() label: string;
  @Prop() invalid: boolean = false;
  @Prop() required: boolean = false;
  @Prop() requiredText: string = '*';
  @Prop() optionalText: string;
  @Prop() invalidText: string;
  @Prop() helperText: string;
  @Prop() helperId: string;
  @Prop() labelTag: string = 'label';
  @Prop() tag: string = 'div';
  @Prop() name: string;
  @Prop() invalidIcon: string = 'error';
  @Prop() labelClass: string;
  @Prop() ariaLabelOptional: string;
  @Prop() ariaLabelRequired: string;

  getClass(): string {
    return [
      `ecl-form-group`,
      this.styleClass
    ].join(' ');
  }

  render() {
    return (
      <this.tag
        class={this.getClass()}
        aria-describedby={this.helperId}
      >
      { this.label ?
        <this.labelTag
          class={`ecl-form-label ${this.invalid ? 'ecl-form-label--invalid' : ''} ${this.labelClass || ''}`}
          {...(this.ariaLabelRequired ? { 'aria-label': this.ariaLabelRequired } : {})}
        >
          {this.label}
      {this.required && this.requiredText ? (
        <span
          class="ecl-form-label__required" 
          {...(this.ariaLabelRequired ? { 'aria-label': this.ariaLabelRequired } : {})}
        >
          {this.requiredText}
        </span>
      ) : ''}
      { !this.required && this.optionalText ?
        <span
          class="ecl-form-label__optional"
          {...(this.ariaLabelOptional ? { 'aria-label': this.ariaLabelOptional } : {})}
        >
          {this.optionalText}</span> : ''
      }
        </this.labelTag> : ''
      }
      { this.helperText ? 
        <div
          class="ecl-help-block"
        >
          {this.helperText}
        </div> : ''
      }
      { this.invalid && this.invalidText ?
        <div class="ecl-feedback-message">
        { this.invalidIcon ? 
          <ecl-icon
            styleClass={`ecl-feedback-message__icon sc-ecl-form-group-${this.theme}`}
            icon={this.invalidIcon}
            size="m"
          ></ecl-icon> : '' 
        }
          {this.invalidText}
        </div> : '' 
      }
        <slot></slot>
      </this.tag>
    )
  }
}