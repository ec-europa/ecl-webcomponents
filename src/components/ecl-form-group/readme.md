# ecl-form-group



<!-- Auto Generated Below -->


## Properties

| Property            | Attribute             | Description | Type      | Default     |
| ------------------- | --------------------- | ----------- | --------- | ----------- |
| `ariaLabelOptional` | `aria-label-optional` |             | `string`  | `undefined` |
| `ariaLabelRequired` | `aria-label-required` |             | `string`  | `undefined` |
| `helperId`          | `helper-id`           |             | `string`  | `undefined` |
| `helperText`        | `helper-text`         |             | `string`  | `undefined` |
| `invalid`           | `invalid`             |             | `boolean` | `false`     |
| `invalidIcon`       | `invalid-icon`        |             | `string`  | `'error'`   |
| `invalidText`       | `invalid-text`        |             | `string`  | `undefined` |
| `label`             | `label`               |             | `string`  | `undefined` |
| `labelClass`        | `label-class`         |             | `string`  | `undefined` |
| `labelTag`          | `label-tag`           |             | `string`  | `'label'`   |
| `name`              | `name`                |             | `string`  | `undefined` |
| `optionalText`      | `optional-text`       |             | `string`  | `undefined` |
| `required`          | `required`            |             | `boolean` | `false`     |
| `requiredText`      | `required-text`       |             | `string`  | `'*'`       |
| `styleClass`        | `style-class`         |             | `string`  | `''`        |
| `tag`               | `tag`                 |             | `string`  | `'div'`     |
| `theme`             | `theme`               |             | `string`  | `'ec'`      |


## Dependencies

### Used by

 - [ecl-search-form](../ecl-search-form)

### Depends on

- [ecl-icon](../ecl-icon)

### Graph
```mermaid
graph TD;
  ecl-form-group --> ecl-icon
  ecl-search-form --> ecl-form-group
  style ecl-form-group fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
