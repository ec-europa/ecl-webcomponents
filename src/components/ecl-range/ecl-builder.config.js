const path = require('path');
const pkg = require('./package.json');

const nodeModules = path.resolve(__dirname, '../../../node_modules');
const outputFolder = './build';

// SCSS includePaths
const includePaths = [nodeModules];

const banner = `${pkg.name} - ${pkg.version} Built on ${new Date().toISOString()}`;

module.exports = {
  scripts: [
    {
      entry: path.resolve(nodeModules, '@ecl/vanilla-component-range/range.js'),
      dest: path.resolve(outputFolder, 'scripts/ecl-range-vanilla.js'),
      options: {
        banner,
        moduleName: 'ECL',
        includePaths,
        sourceMap: false,
      },
    },
  ],
  styles: [
    {
      entry: path.resolve(__dirname, 'src/styles/ecl-range-ec.scss'),
      dest: path.resolve(outputFolder, 'styles/ecl-range-ec.css'),
      options: {
        banner,
        includePaths,
        sourceMap: false,
      },
    },
    {
      entry: path.resolve(__dirname, 'src/styles/ecl-range-eu.scss'),
      dest: path.resolve(outputFolder, 'styles/ecl-range-eu.css'),
      options: {
        banner,
        includePaths,
        sourceMap: false,
      },
    },
  ],
};
