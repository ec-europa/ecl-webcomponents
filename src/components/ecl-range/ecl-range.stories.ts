const getArgs = () => {
  return {
    width: 'm',
    disabled: false,
    required: true,
    invalid: false,
    inputId: 'input-id',
    name: 'input-name',
    max: 30,
    min: 0,
    label: 'Range slider',
    defaultValue: 15,
    step: 1,
  };
};

const getArgTypes = () => {
  return {
    width: {
      type: { name: 'select' },
      options: ['s', 'm', 'l'],
      description: 'Width of the input',
      table: {
        category: 'Styles',
      },
    },
    disabled: {
      type: { name: 'boolean' },
      description: 'Disabled input',
      table: {
        category: 'States',
      },
    },
    required: {
      type: { name: 'boolean' },
      description: 'Required input',
      table: {
        category: 'States',
      },
    },
    invalid: {
      type: { name: 'boolean' },
      description: 'Invalid input',
      table: {
        category: 'States',
      },
    },
    inputId: {
      name: 'Range id',
      type: { name: 'string' },
      description: 'Range variant',
      table: {
        category: 'Content',
      },
    },
    name: {
      type: { name: 'string' },
      description: 'Name attribute',
      table: {
        category: 'Content',
      },
    },
    label: {
      type: { name: 'string' },
      description: 'Form group label',
      table: {
        category: 'Content',
      },
    },
    defaultValue: {
      name: 'value',
      type: 'number',
      description: 'Default value of the input',
      table: {
        category: 'Content',
      },
    },
    min: {
      type: 'number',
      description: 'Minimum value',
      table: {
        type: { summary: 'number' },
        category: 'Range input',
      },
    },
    max: {
      type: 'number',
      description: 'Maximun value',
      table: {
        type: { summary: 'number' },
        category: 'Range input',
      },
    },
    step: {
      type: 'number',
      description: 'Maximun value',
      table: {
        type: { summary: 'number' },
        category: 'Range input',
      },
    },
  };
};

export default {
  title: 'Components/form/range',
};

const Template = args =>
`<ecl-form-group
  label="Range slider"
  required=${args.required}
  optional-text="(optional)"
  helper-text="This is the input helper text"
  invalid=${args.invalid}
  disabled=${args.disabled}
  invalid-text="This is an error message"
  theme="${args.theme}"
  aria-label-optional="Aria label optional"
  aria-label-required="Aria label required"
>
  <ecl-range
    theme="${args.theme}"
    input-id="${args.inputId}"
    default-value="${args.defaultValue}"
    label="${args.label}"
    required=${args.required}
    disabled=${args.disabled}
    invalid=${args.invalid}
    width="${args.width}"
    name="${args.name}"
    min=${args.min}
    max=${args.max}
    ecl-script
    step=${args.step}
    value-label="Value: "
  >

  </ecl-range>
</ecl-form-group>`;

export const Range = Template.bind({});
Range.storyName = 'default';
Range.args = getArgs();
Range.argTypes = getArgTypes();
