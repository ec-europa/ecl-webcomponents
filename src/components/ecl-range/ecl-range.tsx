import { Component, h, Prop, Element, getAssetPath } from '@stencil/core';
declare const ECL: any;

@Component({
  tag: 'ecl-range',
  styleUrls: {
    ec: './build/styles/ecl-range-ec.css',
    eu: './build/styles/ecl-range-eu.css',
  },
  shadow: false,
  scoped: true,
  assetsDirs: ['build'],
})
export class EclRange {
  @Element() el: HTMLElement;
  @Prop() theme: string = 'ec';
  @Prop() styleClass: string;
  @Prop() inputClass: string;
  @Prop() eclScript: boolean = false;
  @Prop() disabled: boolean = false;
  @Prop() required: boolean = false;
  @Prop() invalid: boolean = false;
  @Prop() helperId: string;
  @Prop() helperText: string;
  @Prop() placeholder: string;
  @Prop() width: string = 'm';
  @Prop() label: string;
  @Prop() inputId: string;
  @Prop() name: string;
  @Prop() defaultValue: string;
  @Prop() max: number;
  @Prop() min: number;
  @Prop() step: number = 1;
  @Prop() valueLabel: string;

  getClass(): string {
    const styleClasses = ['ecl-input', this.styleClass];

    return styleClasses.join(' ');
  }

  getInputClasses(): string {
    const inputClasses = ['ecl-range', `ecl-range--${this.width}`];
    
    if (this.inputClass) {
      inputClasses.push(this.inputClass);
    }

    if (this.disabled) {
      inputClasses.push(`ecl-range--disabled`);
    }

    if (this.invalid) {
      inputClasses.push(`ecl-range--invalid`);
    }

    return inputClasses.join(' ');
  }

  componentDidRender() {
    if (this.eclScript) {
      const src = getAssetPath('./build/scripts/ecl-range-vanilla.js');
      if (document.querySelector(`script[src="${src}"]`)) {
        document.querySelector(`script[src="${src}"]`).remove();
      }
      const script = document.createElement('script');
      script.src = src;
      script.onload = () => {
        const formGroup = this.el.closest('.ecl-form-group');
        formGroup.setAttribute('data-ecl-range', 'data-ecl-range');
        formGroup.classList.add(`sc-ecl-range-${this.theme}`);
        const range = new ECL.Range(formGroup);
        range.init();
      };
      document.body.appendChild(script);
    }
  }

  render() {
    const wrapperAttrs = {};
    wrapperAttrs['data-ecl-range'] = true;
  
    const attributes = {
      type: 'range',
      class: this.getInputClasses(),
      id: this.inputId,
      name: this.name,
      required: this.required,
      disabled: this.disabled,
      placeholder: this.placeholder,
      'data-ecl-range-input': true,
      min: this.min,
      max: this.max,
      step: this.step,
      value: this.defaultValue,
    };

    return (
      <div 
        class={this.getClass()}
        {...wrapperAttrs}
      >
        <input {...attributes} />
      
        <div class="ecl-range__value">
          {this.valueLabel}
          <span class="ecl-range__value-current" data-ecl-range-value-current></span>
        </div>
      </div>
    );
  }
}