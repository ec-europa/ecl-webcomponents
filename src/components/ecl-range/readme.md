# ecl-range



<!-- Auto Generated Below -->


## Properties

| Property       | Attribute       | Description | Type      | Default     |
| -------------- | --------------- | ----------- | --------- | ----------- |
| `defaultValue` | `default-value` |             | `string`  | `undefined` |
| `disabled`     | `disabled`      |             | `boolean` | `false`     |
| `eclScript`    | `ecl-script`    |             | `boolean` | `false`     |
| `helperId`     | `helper-id`     |             | `string`  | `undefined` |
| `helperText`   | `helper-text`   |             | `string`  | `undefined` |
| `inputClass`   | `input-class`   |             | `string`  | `undefined` |
| `inputId`      | `input-id`      |             | `string`  | `undefined` |
| `invalid`      | `invalid`       |             | `boolean` | `false`     |
| `label`        | `label`         |             | `string`  | `undefined` |
| `max`          | `max`           |             | `number`  | `undefined` |
| `min`          | `min`           |             | `number`  | `undefined` |
| `name`         | `name`          |             | `string`  | `undefined` |
| `placeholder`  | `placeholder`   |             | `string`  | `undefined` |
| `required`     | `required`      |             | `boolean` | `false`     |
| `step`         | `step`          |             | `number`  | `1`         |
| `styleClass`   | `style-class`   |             | `string`  | `undefined` |
| `theme`        | `theme`         |             | `string`  | `'ec'`      |
| `valueLabel`   | `value-label`   |             | `string`  | `undefined` |
| `width`        | `width`         |             | `string`  | `'m'`       |


----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
