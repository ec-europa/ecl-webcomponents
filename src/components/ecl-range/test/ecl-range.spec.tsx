import { newSpecPage } from '@stencil/core/testing';
import { EclRange } from '../ecl-range';
import { EclFormGroup } from '../../ecl-form-group/ecl-form-group';

describe('ecl-range', () => {
  it('renders', async () => {
    const page = await newSpecPage({
      components: [EclRange, EclFormGroup],
      html: `<ecl-form-group
        label="Label"
        required
        helper-text="This is the input helper text"
        theme="ec"
      >
        <ecl-range
          theme="ec"
          input-id="input-id"
          default-value="15"
          label="Range slider"
          required
          width="m"
          name="input-range"
          min=1
          max=30
          ecl-script
          step=1
          value-label="Value: "
        >
        </ecl-range>
      </ecl-form-group>`,
    });

    expect(page.root).toMatchSnapshot();
  });
});
