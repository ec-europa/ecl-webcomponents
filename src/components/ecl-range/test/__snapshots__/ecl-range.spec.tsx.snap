// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ecl-range renders 1`] = `
<ecl-form-group helper-text="This is the input helper text" label="Label" required="" theme="ec">
  <!---->
  <div class="ecl-form-group">
    <label class="ecl-form-label" for="input-id" id="input-id-label">
      Label
      <span class="ecl-form-label__required">
        *
      </span>
    </label>
    <div class="ecl-help-block" id="input-id-helper">
      This is the input helper text
    </div>
    <ecl-range default-value="15" ecl-script="" input-id="input-id" label="Range slider" max="30" min="1" name="input-range" required="" step="1" theme="ec" value-label="Value: " width="m">
      <div class="ecl-input" data-ecl-range>
        <input class="ecl-range ecl-range--m" data-ecl-range-input id="input-id" max="30" min="1" name="input-range" required="" step="1" type="range" value="15">
        <div class="ecl-range__value">
          Value:
          <span class="ecl-range__value-current" data-ecl-range-value-current></span>
        </div>
      </div>
    </ecl-range>
  </div>
</ecl-form-group>
`;
